### DISCUSSION TITLE
Retrieving Reserved Keywords of the Ballerina Language

### DISCUSSION THREAD

**Discussion by rasika** [ballerina-platform/ballerina-lang#28827] [Ideas]
Is there anyway to expose **LexerTerminals** defined in https://github.com/ballerina-platform/ballerina-lang/blob/master/compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/LexerTerminals.java ? 

This is required for many langserver features such completions, code-actions to get reserved keywords of the Ballerina language.

Key requirement is there's no way to skip suggesting **reserved keywords as variables-names**. not only for the variable names, when suggesting auto-imports we need to escape module-names that has reserved keywords. Even I believe keywords should be transparent for the completion suggestions as well. So that we don't need duplicate and maintain 'em in langserver level.

looping @sameerajayasoma @Kishanthan @hasithaa @lochana-chathura @rdulmina

**Comment by nadeeshaan** [comment]
Currently, we have a major blocker when it comes to the code actions where we generate variable names with the keywords. It is mandatory for the language server to get the keywords and also, as @rasika mentioned, we are keeping a list of preserved keywords (not all) in language server for snippet generation as well. We need to prioritise this

**Comment by lochana-chathura** [comment]
Another requirement from the LS side could be to reuse the identifier matching logic used in the parser. Refer to the comment https://github.com/ballerina-platform/ballerina-lang/pull/28798#discussion_r581635048 by @rasika.

We can do the following from the parser's side to address this. I'm not very sure exposing this way is a good idea.

- We do the identifier matching here in the lexer. https://github.com/ballerina-platform/ballerina-lang/blob/317188576353a17d5abbde7fb1e3bb153811d0ac/compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaLexer.java#L783

- It is not merely an `isIdentifier(String txt)` method. We read the input sequence char by char using the attached CharReader to capture the identifier. So we cannot expose as it is. 

- But, maybe we can simply write a method to boot up a parser and try to parse an identifier from a given string and then return whether it is valid. 

- Also, currently, we don't have a place to expose such API.

**Reply by rasika** [reply]
Would like to add more on this, I understand that the lexer is working on a state base and won't be a good idea to expose/implement `isIdentifier()` there. What we really need is a set of utils **maintained** by parser/compiler module(that make sure upto date parser behavior). This may include `isKeyword(string text)` and `isIdentifier(string text)` same as **roslyn** compiler platform has `SyntaxFacts` class.

**Reply by lochana-chathura** [reply]
The lexer is fed a stream of characters. So it processes each character whether it is valid and keeps increasing the head of the character reader until an end of token character (whitespace char, newline, etc) is reached. 

In order to utilize the identifier processing, we can create a character stream from the text and then fed it to the lexer.
I think it is also faster compared to a regex matching.

eg.
```java
public boolean isIdentifier(String text) {
    CharReader charReader = CharReader.from(text);
    BallerinaLexer lexer = new BallerinaLexer(charReader);
    return text.equals(new TokenReader(lexer).peek().text());
}
```

**Reply by lochana-chathura** [reply]
I am +1 to have some space to expose API. I really have not much idea about what could be the negative consequences of doing so.

**Reply by rasika** [reply]
@sameerajayasoma @Kishanthan @hasithaa Can we get a `SyntaxFacts` implementation through the Compiler APIs for `isKeyword(string text)` and `isIdentifier(string text)` ? Seems like the implementation is straightforward.

**Reply by hasithaa** [reply]
@rasika, Even though `isKeyword` implementation is simple, I think that is not the correct way to handle this use case. Ballerina keywords are contextual, for example, `where` is not a keyword in every context.  

So I would suggest keeping all possible keywords for now on LS side, until we figure out the proper API for this.

**Reply by rasika** [reply]
@hasithaa there's nothing wrong in calculating `where` as a permanent keyword. I think we can forget being contextual for the first phase. WDYT?

**Reply by hasithaa** [reply]
That is correct, But we can't add APIs randomly. We need to design this properly.

**Comment by lochana-chathura** [comment]
For anyone going through this discussion, this requirement was addressed with `SyntaxInfo` API via PR #32806
